### PR TITLE
Handle out-of-bounds negative stack offset without crashing (fixes #1211)

### LINKED ISSUE #1211: Verifier crash (uncaught InternalError) on out-of-bounds stack access: negative offset narrowed to unsigned in array domain
## Summary

Analyzing a program that performs a stack access at a **negative constant offset** (a stack pointer walked below its frame) crashes the verifier with an uncaught `InternalError` instead of cleanly rejecting the out-of-bounds access. This is a robustness / denial-of-service hole: a malformed or adversarial program aborts the analyzer.

```
CRAB ERROR: Number -8 does not fit into m; function narrow, line 208
```

## Root cause

`src/crab/array_domain.cpp` — the strong-update (constant-offset) path narrows the byte offset to the unsigned cell-offset type `Index` (`uint64_t`):

- `kill_and_find_var` (used by `havoc`, `havoc_type`, `store`, `store_type`): `offset_t o(n->narrow<Index>())`
- `split_number_var`: `const offset_t o(n->narrow<Index>())`

When the singleton offset `n` is **negative** (an access below the stack frame), `Number::narrow<Index>()` (`src/arith/num_big.hpp:207`) throws `InternalError` because a negative value does not fit an unsigned type. `CRAB_ERROR` throws rather than being caught and turned into a rejection, so the whole analysis aborts.

Backtrace:
```
Number::narrow<unsigned long>()
kill_and_find_var<...havoc_type...>()
ArrayDomain::havoc_type(...)
EbpfTransformer::do_store_stack(...)
...
InterleavedFwdFixpointIterator::run(...)
```

## Reproduction

A loop wholly inside a bpf-to-bpf callee that walks a stack pointer one byte lower each iteration and writes through it (the callee's own `r10` frame). Under widening the offset stays stack-typed while drifting below the frame, producing a negative constant offset at the store:

```
<sub>:
    r1 = 0
    r2 = r10
    r2 -= 8
<sub_loop>:
    *(u8 *)(r2 + 0) = 0
    r2 -= 1
    r1 += 1
    if r1 < 4096 goto <sub_loop>
    exit
```

(Reaching this specific path depends on callee loops being widened, which is itself gated by #1203; the underlying array-domain crash is independent and is exercised directly by a unit test that calls `store_type` / `havoc_type` / `havoc` / `store` with a negative singleton offset.)

## Expected

An out-of-bounds stack access must be **rejected** by the assertion checker (`Lower bound must be at least r10.stack_offset - subprogram_stack_size`), not crash the analyzer.

## Fix

Guard the constant-offset path with `n->fits<Index>()` (exactly the condition under which the narrow is safe). A negative offset has no corresponding stack cell, so the cell bookkeeping is skipped and the out-of-bounds access is left for the checker to reject. Verified that the crash becomes a clean rejection and that the in-bounds cell is unaffected.

_Found while adding an accept-unsafe regression test for #1203._

### PR DESCRIPTION
**Pull request by elazarg**
## Summary

Fixes #1211 — a robustness/DoS hole where a stack access at a **negative constant offset** (a stack pointer walked below its frame) crashed the verifier with an uncaught `InternalError` instead of rejecting the out-of-bounds access:

```
CRAB ERROR: Number -8 does not fit into m; function narrow, line 208
```

## Root cause

The strong-update (constant-offset) path narrows the byte offset to the unsigned cell-offset type `Index` (`uint64_t`):

- `kill_and_find_var` (used by `havoc`, `havoc_type`, `store`, `store_type`)
- `split_number_var`

A negative singleton offset does not fit an unsigned type, so `Number::narrow<Index>()` throws and aborts the entire analysis rather than letting the assertion checker reject the access.

## Fix

Guard the constant-offset path with `n->fits<Index>()` — exactly the condition under which the narrow is safe. A negative offset has no corresponding stack cell, so the cell bookkeeping is skipped and the out-of-bounds access is left for the checker to reject.

Verified end-to-end (with #1203's callee-loop widening applied, which is how the negative offset arises in a real program) that the crash becomes a **clean rejection** — `Lower bound must be at least r10.stack_offset - subprogram_stack_size` — and **not** a silent accept.

## Test

`test_array_domain.cpp`: drives `store_type` / `havoc_type` / `havoc` / `store` with a negative singleton offset and asserts no throw and that the in-bounds cell is untouched. It throws the `InternalError` without the fix and passes with it. Full suite: 636 passed, 1 skipped, 0 failures.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01QRunjNUunuZjp6KiBSZXzt

### CODE REVIEW

File: src/crab/array_domain.cpp [hunk 1]
```diff
@@ -328,7 +330,13 @@ static std::optional<std::pair<offset_t, unsigned>> kill_and_find_var(StackCellR
 
     offset_map_t& offset_map = cells.get(kind);
     std::vector<Cell> overlaps;
-    if (const std::optional<Number> n = ii.singleton()) {
+    if (const std::optional<Number> n = ii.singleton(); n && n->fits<Index>()) {
```
[the remark this answers is not in the document]
Comment (elazarg) [reply src/crab/array_domain.cpp]: Good catch — fixed in 9b3da88.

You're right: the first commit only stopped the narrow crash. A negative singleton offset then left `res` empty and fell through to the symbolic kill, whose inclusive range `ii | (ii + elem_size)` (e.g. `[-8, 0]`) shares the endpoint `0` with the tracked cell `[0, 8)`, so a havoc/store at a negative offset erased the offset-0 value. Sound (the OOB access is rejected regardless), but an avoidable precision loss.

Now a negative/out-of-range **constant** offset returns early instead of falling through — the access is entirely below the frame, there's no cell to kill, and the checker rejects it. A constant offset with a *non-singleton width* still uses the symbolic path as before.

I also extended the regression test per your note: it stores a known value in the in-bounds cell, runs an out-of-bounds negative havoc/store, and asserts the value survives. That assertion **fails** under the fall-through version (`{?} == 42`) and passes now.
